contracts/Bonuses.sol

```diff
@@ -0,0 +1,148 @@
+/// @title Bonuses
+/// @dev Aids transferring bonus tokens related to the token sale
+contract Bonuses is Processables {
+    using SafeMath for uint256;
+
+	// named "processables" to aid JS interaction with this and other contracts that inherit from Processables
+	mapping(address => Processable) public processables;
+
+	// Amount of gas sufficient to conduct a single process loop iteration and post-loop operations
+	uint256 public constant PROCESS_GAS_MINIMUM = 100000;
+
+	// Total bonuses amount remaining to process after locking
+	uint256 public remainingTotalBonuses = 0;
+
+	struct Processable {
+		uint256 amount;
+		bool processed;
+	}
+
+	SimpleToken public simpleToken;
+
+	event BonusAdded(address indexed _address, uint256 _amount);
+	event BonusProcessed(address indexed _address, uint256 _amount, uint256 _index);
+
+	// limits execution to SimpleToken.owner
+	modifier onlySTOwner() {
+        require(msg.sender == simpleToken.owner());
+        _;
+    }
+
+
+	function Bonuses(SimpleToken _simpleToken)
+			Processables()
+			public
+	{
+		require(address(_simpleToken) != address(0));
+		simpleToken = _simpleToken;
+	}
+
+	/// @dev Adds bonus information to processables
+	/// @param _address address
+	/// @param _amount amount
+	/// @return true
+	function add(address _address, uint256 _amount)
+			public
+			onlyOwner
+			onlyIfUnlocked
+			returns (bool result)
+	{
+		require(_address != address(0));
+		require(_amount > 0);
+
+		Processable storage processable = processables[_address];
+
+		require(processable.amount == 0);
+		require(add(_address));
+
+		processable.amount = _amount;
+
+		BonusAdded(_address, _amount);
+
+		return true;
+	}
+
+	/// @dev Locks this contract
+	function lock()
+			public
+			returns (bool result)
+	{
+		// onlyOwner
+		result = super.lock();
+
+		for (uint256 i = 0; i < addresses.length; i++)
+			remainingTotalBonuses = remainingTotalBonuses.add(processables[addresses[i]].amount);
+	}
+
+	/// @dev Approves this contract
+	function approve()
+			external
+			onlySTOwner
+			returns (bool result)
+	{
+		return approveInternal();
+	}
+
+	/// @dev Disapproves this contract
+	function disapprove()
+			external
+			onlySTOwner
+			returns (bool result)
+	{
+		return disapproveInternal();
+	}
+
+	/// @dev Processes as many bonuses as possible from a certain index
+	/// @param _from from
+	/// @return true if processes last bonus, false if otherwise
+	function process(uint256 _from)
+			public
+			onlyOwner
+			onlyIfApproved
+			returns (bool result)
+	{
+		// Confirm that this contract is approved to transfer a sufficient amount of ST
+		address owner = simpleToken.owner();		
+		require(simpleToken.allowance(owner, address(this)) >= remainingTotalBonuses);
+
+		uint256 startingGas = msg.gas;
+
+		// Breaks if there is not enough gas for both a complete iteration and post-loop operations
+		for (uint256 i = _from; i < addresses.length && (block.gaslimit - (startingGas - msg.gas) > PROCESS_GAS_MINIMUM); i++) {
```

Comment:
Not sure I understand the condition in this loop. The intent is to make sure we don't run out of gas or exceed the block gas limit? With the current check let's say block.gaslimit is 4.7M and starting gas is 2M then you might end up with 0 gas left and your condition is still 4.7 - (2M - 0) = 2.7M > PROCESS_GAS_MINIMUM and it will continue looping and throw out of gas. Wondering whether just checking msg.gas > PROCESS_GAS_MINIMUM would be sufficient.

Reply:
Ah, you're right—though that was exactly what I was trying to avoid. Checking `msg.gas` alone is not sufficient because msg.gas could be higher than the block.gas. So I should add that `msg.gas > PROCESS_GAS_MINIMUM` as well. Confirmed this with tests—will push up.

Reply:
Actually: (1) my test coverage is insufficient and (2) this logic needs to be reworked 😰 . Now for some actual TDD.